flocker/node/agents/test/test_blockdevice.py

```diff
@@ -3395,23 +3478,99 @@ class CreateBlockDeviceDatasetInterfaceTests(
     """
 
 
-class CreateBlockDeviceDatasetImplementationTests(SynchronousTestCase):
+class CreateBlockDeviceDatasetImplementationMixin(object):
     """
-    ``CreateBlockDeviceDataset`` implementation tests.
+    Utility Mixin for ``CreateBlockDeviceDataset`` implementation tests.
     """
-    def setUp(self):
-        self.api = loopbackblockdeviceapi_for_test(
-            self,
-            allocation_unit=LOOPBACK_ALLOCATION_UNIT
+
+    def _create_blockdevice_dataset(self, dataset_id, maximum_size,
+                                    metadata=pmap({})):
+        """
+        Call ``CreateBlockDeviceDataset.run`` with a ``BlockDeviceDeployer``.
+
+        :param UUID dataset_id: The uuid4 identifier for the dataset which will
+            be created.
+        :param int maximum_size: The size, in bytes, of the dataset which will
+            be created.
+        :param pmap(unicode, unicode) metadata: The metadata for the dataset.
+
+        :returns: A 3-tuple of:
+            * ``BlockDeviceVolume`` created by the run operation
+            * The ``FilePath`` of the device where the volume is attached.
+            * The ``FilePath`` where the volume is expected to be mounted.
+        """
+        expected_mountpoint = self.mountroot.child(
+            unicode(dataset_id).encode("ascii")
         )
-        self.mountroot = mountroot_for_test(self)
-        self.deployer = BlockDeviceDeployer(
-            node_uuid=uuid4(),
-            hostname=u"192.0.2.10",
-            block_device_api=self.api,
-            mountroot=self.mountroot
+
+        dataset = Dataset(
+            dataset_id=unicode(dataset_id),
+            maximum_size=maximum_size,
+            metadata=metadata
+        )
+
+        change = CreateBlockDeviceDataset(
+            dataset=dataset, mountpoint=expected_mountpoint
         )
 
+        run_state_change(change, self.deployer)
+
+        [volume] = self.api.list_volumes()
+        device_path = self.api.get_device_path(volume.blockdevice_id)
+        return (
+            volume, device_path, expected_mountpoint,
+            self.api.compute_instance_id()
+        )
+
+
+def make_createblockdevicedataset_mixin(profiled_api):
+    """
+    Constructs a base class for tests that verify the implementation of
+    ``CreateBlockDeviceDataset``.
+
+    This ``IStateChange`` needs to be tested in two configurations:
```

Comment:
Hm.  Alternatively, there should be two different `IStateChange` implementations...  Consider this for follow-up work (maybe add a note to the issue for breaking `CreateBlockDeviceDataset` up into pieces).

Ideally, there shouldn't be branches in an `IStateChange.run` method.  Decisions should all have been made by the time the object was created.